pkg/endpoint/events.go

```diff
@@ -0,0 +1,104 @@
+// Copyright 2019 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package endpoint
+
+import (
+	"sync"
+
+	"github.com/cilium/cilium/pkg/eventqueue"
+)
+
+// EndpointRegenerationEvent contains all fields necessary to regenerate an endpoint.
+type EndpointRegenerationEvent struct {
+	owner        Owner
+	regenContext *regenerationContext
+	ep           *Endpoint
+}
+
+// Handle handles the regeneration event for the endpoint.
+func (ev *EndpointRegenerationEvent) Handle(res chan interface{}) {
+	e := ev.ep
+	owner := ev.owner
+	regenContext := ev.regenContext
+
+	err := e.RLockAlive()
+	if err != nil {
+		e.LogDisconnectedMutexAction(err, "before regeneration")
+		res <- &EndpointRegenerationResult{
+			err: err,
+		}
+
+		return
+	}
+	e.RUnlock()
+
+	// We should only queue the request after we use all the endpoint's
+	// lock/unlock. Otherwise this can get a deadlock if the endpoint is
+	// being deleted at the same time. More info PR-1777.
+	doneFunc := owner.QueueEndpointBuild(uint64(e.ID))
+	if doneFunc != nil {
+		e.getLogger().Debug("Dequeued endpoint from build queue")
+
+		regenContext.DoneFunc = doneFunc
+
+		err = ev.ep.regenerate(ev.owner, ev.regenContext)
+
+		doneFunc()
+		e.notifyEndpointRegeneration(owner, err)
+	} else {
+		e.getLogger().Debug("My request was cancelled because I'm already in line")
+	}
+
+	res <- &EndpointRegenerationResult{
+		err: err,
+	}
+	return
+}
+
+// EndpointRegenerationResult contains the results of an endpoint regeneration.
+type EndpointRegenerationResult struct {
+	err error
+}
+
+// EndpointRevisionBumpEvent contains all fields necessary to bump the policy
+// revision of a given endpoint.
+type EndpointRevisionBumpEvent struct {
+	Rev uint64
+	ep  *Endpoint
+}
+
+// Handle handles the revision bump event for the Endpoint.
+func (ev *EndpointRevisionBumpEvent) Handle(res chan interface{}) {
+	// TODO: if the endpoint is not in a 'ready' state that means that
```

Comment:
What's the current impact of this TODO?

Reply:
(And to answer your question, yeah I would think that if the event is impossible to achieve then the error should be signalled.. at the moment, it would appear it's being disregarded, but we could as easily at least log it and see whether it crops up anywhere)

Reply:
I put this TODO there to get feedback from reviewers - so thanks for taking a look :) The policy revision was only incremented previously if the regeneration was successful, so this was meant to make this more explicit in the function itself. 

I'll add a log.

Reply:
I added the state-checking functionality here as the last commit. This encapsulates both of the cases in which the endpoint's policy revision should be incremented.